pkg/nodeidentity/azure/client.go

```diff
@@ -22,81 +22,97 @@ import (
 	"fmt"
 	"io"
 	"net/http"
+	"strings"
 
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
 	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
 	compute "github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute"
-	"k8s.io/kops/upup/pkg/fi"
 )
 
-type instanceComputeMetadata struct {
-	ResourceGroupName string `json:"resourceGroupName"`
-	SubscriptionID    string `json:"subscriptionId"`
-}
-
-type instanceMetadata struct {
-	Compute *instanceComputeMetadata `json:"compute"`
-}
-
 // client is an Azure client.
 type client struct {
-	metadata     *instanceMetadata
-	vmssesClient *compute.VirtualMachineScaleSetsClient
+	subscriptionID string
+	vmClient       *compute.VirtualMachinesClient
+	vmssClient     *compute.VirtualMachineScaleSetVMsClient
 }
 
 // newClient returns a new Client.
 func newClient() (*client, error) {
-	m, err := queryInstanceMetadata()
+	metadata, err := queryComputeInstanceMetadata()
 	if err != nil {
 		return nil, fmt.Errorf("error querying instance metadata: %s", err)
 	}
-	if m.Compute.SubscriptionID == "" {
-		return nil, fmt.Errorf("empty subscription name")
-	}
-	if m.Compute.ResourceGroupName == "" {
-		return nil, fmt.Errorf("empty resource group name")
+	if metadata.SubscriptionID == "" {
+		return nil, fmt.Errorf("empty subscription ID")
 	}
 
 	cred, err := azidentity.NewDefaultAzureCredential(nil)
 	if err != nil {
 		return nil, fmt.Errorf("creating identity: %w", err)
 	}
 
-	vmssesClient, err := compute.NewVirtualMachineScaleSetsClient(m.Compute.SubscriptionID, cred, nil)
+	vmClient, err := compute.NewVirtualMachinesClient(metadata.SubscriptionID, cred, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating VMs client: %w", err)
+	}
+
+	vmssClient, err := compute.NewVirtualMachineScaleSetVMsClient(metadata.SubscriptionID, cred, nil)
 	if err != nil {
-		return nil, fmt.Errorf("creating VMSS client: %w", err)
+		return nil, fmt.Errorf("creating VMSS VMs client: %w", err)
 	}
 
 	return &client{
-		metadata:     m,
-		vmssesClient: vmssesClient,
+		vmClient:   vmClient,
+		vmssClient: vmssClient,
 	}, nil
 }
 
-// getVMScaleSet returns the specified VM ScaleSet.
-func (c *client) getVMScaleSet(ctx context.Context, vmssName string) (*compute.VirtualMachineScaleSet, error) {
-	opts := &compute.VirtualMachineScaleSetsClientGetOptions{
-		Expand: fi.PtrTo(compute.ExpandTypesForGetVMScaleSetsUserData),
+func (c *client) getVMTags(ctx context.Context, providerID string) (map[string]*string, error) {
+	if !strings.HasPrefix(providerID, "azure://") {
```

Comment:
Nit: I would have put `"azure://"` in a const given the multiple occurrences.

Reply:
Needs to be done for all providers, so I prefer to do it separately.